x-pack/plugins/ml/server/lib/alerts/register_anomaly_detection_alert_type.ts

```diff
@@ -42,10 +62,88 @@ export type AnomalyDetectionAlertContext = AnomalyDetectionAlertBaseContext & {
   topInfluencers?: InfluencerAnomalyAlertDoc[];
 };
 
+export type ExecutorOptions<P extends RuleTypeParams> = RuleExecutorOptions<
+  P,
+  RuleTypeState,
+  {},
+  AnomalyDetectionAlertContext,
+  typeof ANOMALY_SCORE_MATCH_GROUP_ID,
+  MlAnomalyDetectionAlert
+>;
+
 export const ANOMALY_SCORE_MATCH_GROUP_ID = 'anomaly_score_match';
 
 export type AnomalyScoreMatchGroupId = typeof ANOMALY_SCORE_MATCH_GROUP_ID;
 
+export const ANOMALY_DETECTION_AAD_INDEX_NAME = 'ml.anomaly-detection';
+
+const ML_ALERT_NAMESPACE = ALERT_NAMESPACE;
+
+export const ALERT_ANOMALY_DETECTION_JOB_ID = `${ML_ALERT_NAMESPACE}.job_id` as const;
+
+export const ALERT_ANOMALY_SCORE = `${ML_ALERT_NAMESPACE}.anomaly_score` as const;
+export const ALERT_ANOMALY_IS_INTERIM = `${ML_ALERT_NAMESPACE}.is_interim` as const;
+export const ALERT_ANOMALY_TIMESTAMP = `${ML_ALERT_NAMESPACE}.anomaly_timestamp` as const;
+
+export const ALERT_TOP_RECORDS = `${ML_ALERT_NAMESPACE}.top_records` as const;
+export const ALERT_TOP_INFLUENCERS = `${ML_ALERT_NAMESPACE}.top_influencers` as const;
+
+export const ANOMALY_DETECTION_AAD_CONFIG: IRuleTypeAlerts<MlAnomalyDetectionAlert> = {
+  context: ANOMALY_DETECTION_AAD_INDEX_NAME,
+  mappings: {
+    fieldMap: {
+      [ALERT_ANOMALY_DETECTION_JOB_ID]: {
+        type: ES_FIELD_TYPES.KEYWORD,
+        array: false,
+        required: true,
+      },
+      [ALERT_ANOMALY_SCORE]: { type: ES_FIELD_TYPES.DOUBLE, array: false, required: false },
+      [ALERT_ANOMALY_IS_INTERIM]: { type: ES_FIELD_TYPES.BOOLEAN, array: false, required: false },
+      [ALERT_ANOMALY_TIMESTAMP]: { type: ES_FIELD_TYPES.DATE, array: false, required: false },
+      [ALERT_TOP_RECORDS]: {
+        type: ES_FIELD_TYPES.NESTED,
```

Comment:
We discussed it with @peteharverson and agreed it might be useful to fetch alerts with certain influencers, e.g. 
```
GET .alerts-ml.anomaly-detection.alerts-default/_search
{
  "query": {
    "nested": {
      "path": "kibana.alert.top_influencers",
      "query": {
        "bool": {
          "must": [
            {
              "match": {
                "kibana.alert.top_influencers.influencer_field_name": "key"
              }
            }
          ]
        }
      }
    }
  }
}
```
If I'm not mistaken it's only possible with the nested type.

Reply:
I agree with @droberts195 that we need to be especially cautious with using the `nested` type. There are tons of pitfalls with the nested field type per the official docs: https://www.elastic.co/guide/en/elasticsearch/reference/current/nested.html. Additionally, [Kibana does not handle  nested fields well](https://github.com/elastic/kibana/issues/1084), so this will make consuming the data stored in these nested field types complicated as well.

If we use the `object` type, we'd still be able to perform the query to match the influencers. The biggest limitation is that we can't query for multiple fields that appear within the array because of the way ES flattens arrays: https://www.elastic.co/guide/en/elasticsearch/reference/current/nested.html#nested-arrays-flattening-objects

Reply:
We're running into general limitations of Elasticsearch and Kibana functionality.

"Show me ML alerts where one of the influencer fields was `airline`" will be possible with an `object` mapping.

But, "Show me ML alerts where one of the influencer fields was `airline` and its influencer score was greater than 90" will not be reliable with `object`. This is because `influence_field_name: airline` and `influencer_score > 90` might match on different array elements. So it might be that `destination_airport` was the influencer with the score above 90, not `airline`.

Given the difficulties of working with `nested` in Kibana, and other drawbacks with `flattened`, it sounds like `object` is the mapping to go for here. But this _will_ restrict the search functionality that can be built on top of these alerts to some extent. If these limitations are likely to cause problems then we need to hold this functionality back for a later release and think through a more major redesign of the document structure.

Reply:
Many thanks for the feedback and suggestions @droberts195, @kobelb. Updated to the `object` type in https://github.com/elastic/kibana/pull/166349/commits/33d037c43aa592058a9704875dc9ca4ac0b77b46.

Support for queries like this should be sufficient, cc @peteharverson 
```
GET .alerts-ml.anomaly-detection.alerts-default/_search
{
  "query": {
    "term": {
      "kibana.alert.top_influencers.influencer_field_name": {
        "value": "key"
      }
    }
  }
}
```

Reply:
Thanks Dima. We discussed this a bit and came to the conclusion that `object` was the best option of the 3, even though none of the options are perfect. So what you've done tallies with the conclusion we came to.